### PR TITLE
Novel uniform training config 

### PR DESCRIPTION
**Pull request by tuananhroman**
New training config layout - contains every training-related parameter and is stored with each rl agent.
Additionally, a file converter which takes the old hyperparameter.json and converts it into the newly conceptualized format was implemented.
Moreover, reward function loading scheme was simplified.

### CODE REVIEW

File: rl_utils/rl_utils/utils/reward.py [hunk 1]
```diff
@@ -37,11 +36,15 @@ def __init__(
         self._curr_dist_to_path = None
         self.safe_dist = safe_dist
 
-        self.custom_rew_dict = custom_rew_dict
+        # load custom reward function from yaml
+        import tools.train_agent_utils as train_agent_utils
```
Comment (ReykCS) [inline rl_utils/rl_utils/utils/reward.py:40]: Imports ganz oben in die File

File: training/configs/hyperparameters/default.json [hunk 2]
```diff
@@ -20,6 +20,5 @@
   "m_batch_size": 20,
   "n_epochs": 3,
   "clip_range": 0.22,
-  "space_encoder": null,
-  "custom_rew_file": "rule_00"
+  "space_encoder": null
 }
```
Comment (ReykCS) [inline training/configs/hyperparameters/default.json:24]: Die file kann dann raus oder

File: training/configs/training_config.yaml [hunk 3]
```diff
@@ -1,7 +1,8 @@
 ### General
 # in debug_mode no agent directories will be created and no models will be saved
-# further no wandb logging and fake (simulated) multiprocessing for simpler debugging 
-debug_mode: false
+# further no wandb logging and fake (simulated) multiprocessing for traceback
+debug_mode: true
+robot: "jackal"
```
Comment (ReykCS) [inline training/configs/training_config.yaml:5]: wird der nicht in start_training.launch angegeben?

File: training/configs/training_config.yaml [hunk 4]
```diff
@@ -23,36 +24,54 @@ task_mode: "staged"
...
+    curr_stage: 1
+    threshold_type: "succ"
+    upper_threshold: 0.8
+    lower_threshold: 0.6
+
+  ### Stop Training on Threshold
+  # stops training when last stage reached and threshold satisfied
+  stop_training:
+    threshold_type: "succ"
+    threshold: 0.9
```
Comment (ReykCS) [inline training/configs/training_config.yaml:54]: Würde das nicht unter callbacks packen sondern so machen die vorher

File: training/configs/training_config.yaml [hunk 5]
```diff
@@ -23,36 +24,54 @@ task_mode: "staged"
...
-architecture_name: "AGENT_25"
-# path to latest checkpoint; if provided the training will be resumed from that checkpoint
-resume: "jackal_AGENT_25_RobotSpecificEncoder_2023_01_21__03_01"
-reward_function_file: null # "rcppo.yaml"
\ No newline at end of file
+rl_agent:
+  # name of architecture defined in the Policy factory
+  architecture_name: "AGENT_25"
+  # path to latest checkpoint; if provided the training will be resumed from that checkpoint
+  resume: "jackal_AGENT_25_RobotSpecificEncoder_2023_01_21__03_01"
```
Comment (ReykCS) [inline training/configs/training_config.yaml:61]: Das hier würd ich nicht in die Config packen sondern weiterhin als Arg passen. 
Generell sind hier viele Sachen drin die nicht wirklich das training configurieren sondern eher das Setup. Zb usewandb oder n_envs. Es wäre sinnvoll das weiterhin als arg zu haben damit man das schneller ändern kann.

File: training/scripts/train_agent.py [hunk 6]
```diff
@@ -19,10 +19,10 @@ def main():
         rospy.init_node("debug_node", disable_signals=False)
 
     # generate agent name and model specific paths
-    AGENT_NAME = get_agent_name(config)
-    PATHS = get_paths(AGENT_NAME, config)
+    get_agent_name(config)
```
Comment (ReykCS) [inline training/scripts/train_agent.py:22]: Was macht get_agent_name? Wenn du funktion nichts zurück gibt ist "get" hier verwirrend

File: training/scripts/train_agent.py [hunk 7]
```diff
@@ -31,19 +31,18 @@ def main():
     wait_for_nodes(with_ns=ns_for_nodes, n_envs=config["n_envs"], timeout=5)
 
     # initialize hyperparameters (save to/ load from json)
-    params = initialize_hyperparameters(
+    config = initialize_hyperparameters(
```
Comment (ReykCS) [inline training/scripts/train_agent.py:34]: initialize_configs und nicht hyperparams oder?

File: training/tools/hyperparam_file_converter.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+import argparse
```
Comment (ReykCS) [inline training/tools/hyperparam_file_converter.py:1]: hyperparam_file_converter oder config_file_converter?

File: training/tools/hyperparam_file_converter.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+def main():
+    parsed_args = get_parsed_args()
+
+    hyperparams, config = load_files(parsed_args.path)
+
+    config = transfer_entries(hyperparams, config)
+
+    with open(extract_yaml_file_path(parsed_args.path), "w") as outfile:
+        yaml.dump(config, outfile, default_flow_style=False)
+
```
Comment (ReykCS) [inline training/tools/hyperparam_file_converter.py:86]: Hast du das schon für alle alten agents gemacht?

File: training/tools/train_agent_utils.py [hunk 10]
```diff
@@ -101,45 +99,36 @@ def initialize_hyperparameters(
...
-    check_batch_size(n_envs, hyperparams["batch_size"], hyperparams["m_batch_size"])
-    hyperparams["n_steps"] = int(hyperparams["batch_size"] / n_envs)
+    check_batch_size(
+        n_envs,
+        config["rl_agent"]["ppo"]["batch_size"],
+        config["rl_agent"]["ppo"]["m_batch_size"],
+    )
+    config["rl_agent"]["ppo"]["n_steps"] = int(
+        config["rl_agent"]["ppo"]["batch_size"] / n_envs
+    )
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:114]: entpack doch die configs in eine neue variable dann musst du nicht so viel schreiben und ist übersichtlicher:

`params = config["rl_agent"]["ppo"]`

File: training/tools/train_agent_utils.py [hunk 11]
```diff
@@ -101,45 +99,36 @@ def initialize_hyperparameters(
...
+
     if not debug_mode:
-        write_hyperparameters_json(hyperparams, PATHS)
-    print_hyperparameters(hyperparams)
+        write_hyperparameters_yaml(config, PATHS)
+    print_hyperparameters(config["rl_agent"]["ppo"])
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:121]: Wenn das überall hyperparams heisst warum dann in der File ppo? Nenn es doch in der config file auch hyperparams, das wäre konsistenter und leichter zu verstehen

File: training/tools/train_agent_utils.py [hunk 12]
```diff
@@ -101,45 +99,36 @@ def initialize_hyperparameters(
...
+    """
+    Write training_config.yaml to agent directory
 
-def init_params_loaded_agent(PATHS) -> dict:
-    return load_hyperparameters_json(PATHS=PATHS)
+    :param hyperparams: dict containing model specific hyperparameters
+    :param PATHS: dictionary containing model specific paths
+    """
+    with open(PATHS["hyperparams"], "w") as outfile:
+        yaml.dump(hyperparams, outfile, default_flow_style=False)
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:134]: Verstehe nicht, warum hier nur die hyperparams geschrieben werden und nicht gleich die ganze trainings config

File: training/tools/train_agent_utils.py [hunk 13]
```diff
@@ -248,32 +235,32 @@ def update_hyperparam_model(
...
+        model.gamma = params["rl_agent"]["ppo"]["gamma"]
+    if model.n_steps != params["rl_agent"]["ppo"]["n_steps"]:
+        model.n_steps = params["rl_agent"]["ppo"]["n_steps"]
+    if model.ent_coef != params["rl_agent"]["ppo"]["ent_coef"]:
+        model.ent_coef = params["rl_agent"]["ppo"]["ent_coef"]
+    if model.learning_rate != params["rl_agent"]["ppo"]["learning_rate"]:
+        model.learning_rate = params["rl_agent"]["ppo"]["learning_rate"]
+    if model.vf_coef != params["rl_agent"]["ppo"]["vf_coef"]:
+        model.vf_coef = params["rl_agent"]["ppo"]["vf_coef"]
+    if model.max_grad_norm != params["rl_agent"]["ppo"]["max_grad_norm"]:
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:253]: Hier würde sich das entpacken vor den ganzen ifs auch lohnen

File: training/tools/train_agent_utils.py [hunk 14]
```diff
@@ -248,32 +235,32 @@ def update_hyperparam_model(
...
-    if model.vf_coef != params["vf_coef"]:
-        model.vf_coef = params["vf_coef"]
-    if model.max_grad_norm != params["max_grad_norm"]:
-        model.max_grad_norm = params["max_grad_norm"]
-    if model.gae_lambda != params["gae_lambda"]:
-        model.gae_lambda = params["gae_lambda"]
-    if model.n_epochs != params["n_epochs"]:
-        model.n_epochs = params["n_epochs"]
+    if model.batch_size != params["rl_agent"]["ppo"]["batch_size"]:
+        model.batch_size = params["rl_agent"]["ppo"]["batch_size"]
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:242]: Die If statements können eig auch alle weg weil jedes eig folgendes macht

Wenn A nicht gleich B ist, dann setze A = B. Wenn der Code im If also nicht ausgeführt wird ist A = B, dann würde es auch nicht schaden nochmal A = B zu setzen. Das zu entfernen dient also enorm der leserlichkeit und ändert nichts am outcome

File: training/tools/train_agent_utils.py [hunk 15]
```diff
@@ -453,30 +443,28 @@ def _init() -> Union[gym.Env, gym.Wrapper]:
...
                     eval_ns,
-                    params["reward_fnc"],
-                    params["discrete_action_space"],
-                    goal_radius=params["goal_radius"],
-                    max_steps_per_episode=params["eval_max_steps_per_episode"],
-                    task_mode=params["task_mode"],
-                    curr_stage=params["curr_stage"],
+                    config["rl_agent"]["reward_fnc"],
+                    config["rl_agent"]["discrete_action_space"],
+                    oal_radius=config["goal_radius"],
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:464]: goal_radius

File: training/tools/train_agent_utils.py [hunk 16]
```diff
@@ -633,39 +613,38 @@ def init_envs(
...
 
     # threshold settings for training curriculum
     # type can be either 'succ' or 'rew'
     trainstage_cb = arena_cb.InitiateNewTrainStage(
         n_envs=config["n_envs"],
-        treshhold_type=config["training_curriculum"]["threshold_type"],
-        upper_threshold=config["training_curriculum"]["upper_threshold"],
-        lower_threshold=config["training_curriculum"]["lower_threshold"],
-        task_mode=params["task_mode"],
+        treshhold_type=config["callbacks"]["training_curriculum"]["threshold_type"],
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py:640]: hier könnte man auch zumindest die callbacks entpacken

File: training/tools/train_agent_utils.py [hunk 17]
```diff
@@ -689,32 +668,30 @@ def init_callbacks(
...
+                n_steps=config["rl_agent"]["ppo"]["n_steps"],
+                ent_coef=config["rl_agent"]["ppo"]["ent_coef"],
+                learning_rate=config["rl_agent"]["ppo"]["learning_rate"],
+                vf_coef=config["rl_agent"]["ppo"]["vf_coef"],
+                max_grad_norm=config["rl_agent"]["ppo"]["max_grad_norm"],
+                gae_lambda=config["rl_agent"]["ppo"]["gae_lambda"],
+                batch_size=config["rl_agent"]["ppo"]["m_batch_size"],
+                n_epochs=config["rl_agent"]["ppo"]["n_epochs"],
+                clip_range=config["rl_agent"]["ppo"]["clip_range"],
+                tensorboard_log=PATHS["tb"],
```
Comment (ReykCS) [inline training/tools/train_agent_utils.py]: HIer auch entpacken

File: rl_utils/rl_utils/utils/reward.py (hunk 1 above)
Comment (tuananhroman) [reply rl_utils/rl_utils/utils/reward.py:40]: Circular import

File: training/configs/hyperparameters/default.json (hunk 2 above)
Comment (tuananhroman) [reply training/configs/hyperparameters/default.json:24]: Geht raus

File: training/configs/training_config.yaml (hunk 3 above)
Comment (tuananhroman) [reply training/configs/training_config.yaml:5]: Sollte eigentlich nicht

File: training/configs/training_config.yaml (hunk 4 above)
Comment (tuananhroman) [reply training/configs/training_config.yaml:54]: Passt von der Aufteilung her zum Code, aber kann auch so machen wie du meinst. Habs jetzt aber erstmal so integriert, sollte man so lassen können, wenn's nur Kosmetik ist

File: training/scripts/train_agent.py (hunk 7 above)
Comment (tuananhroman) [reply training/scripts/train_agent.py:34]: Änder ich!

File: training/tools/hyperparam_file_converter.py (hunk 8 above)
Comment (tuananhroman) [reply training/tools/hyperparam_file_converter.py:1]: Änder ich!

File: training/tools/hyperparam_file_converter.py (hunk 9 above)
Comment (tuananhroman) [reply training/tools/hyperparam_file_converter.py:86]: Nein, kann der User dann machen, in dem er das Script ausführt

File: training/tools/train_agent_utils.py (hunk 17 above)
Comment (tuananhroman) [reply training/tools/train_agent_utils.py]: Safe

File: training/tools/train_agent_utils.py (hunk 15 above)
Comment (tuananhroman) [reply training/tools/train_agent_utils.py:464]: Danke!

File: training/configs/training_config.yaml (hunk 3 above)
Comment (tuananhroman) [reply training/configs/training_config.yaml:5]: Verändert nichts auf der Simulationsseite

File: training/scripts/train_agent.py (hunk 7 above)
Comment (tuananhroman) [reply training/scripts/train_agent.py:34]: Ist mittlerweile äquivalent

File: training/tools/train_agent_utils.py (hunk 12 above)
Comment (tuananhroman) [reply training/tools/train_agent_utils.py:134]: Änder ich!

File: training/tools/train_agent_utils.py (hunk 14 above)
Comment (tuananhroman) [reply training/tools/train_agent_utils.py]: Ironischerweise habe ich die If-Bedingungen zwecks Leserlichkeit hinzugefügt haha

Comment (tuananhroman) [reply training/tools/train_agent_utils.py]: Änder ich gerne

File: rl_utils/rl_utils/utils/reward.py (hunk 1 above)
Comment (ReykCS) [reply rl_utils/rl_utils/utils/reward.py:40]: Man könnte die Train agent utils file noch ein bisschen auseinander nehmen und einzelne files erstellen. Bietet sich sowieso an weil eine 850 Zeilen Datei doch eher unübersichtlich ist.

File: training/configs/training_config.yaml (hunk 3 above)
Comment (ReykCS) [reply training/configs/training_config.yaml:5]: Was macht dann das Argument model in der start_training.launch

File: training/tools/env_utils.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+            )
+    return env, eval_env
+
+
+def init_envs(
+    config: dict,
+    paths: dict,
+    ns_for_nodes: bool,
+) -> Tuple[VecEnv, VecEnv]:
+    import stable_baselines3.common.vec_env as sb3_env
```
Comment (ReykCS) [inline training/tools/env_utils.py:107]: Du hast oben schon einen import benutz den doch

File: training/tools/general.py [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+    """
+    Write hyperparameters.json to agent directory
+
+    :param hyperparams: dict containing model specific hyperparameters
+    :param PATHS: dictionary containing model specific paths
+    """
+    doc_location = os.path.join(PATHS.get("model"), "hyperparameters.json")
+
+    with open(doc_location, "w", encoding="utf-8") as target:
+        json.dump(hyperparams, target, ensure_ascii=False, indent=4)
```
Comment (ReykCS) [inline training/tools/general.py:88]: Wird das noch gebraucht?

File: training/tools/hyperparam_file_converter.py (hunk 8 above)
Comment (ReykCS) [reply training/tools/hyperparam_file_converter.py:1]: hast du das jetzt geändert?

File: rl_utils/rl_utils/utils/reward.py [hunk 20]
```diff
@@ -37,11 +35,15 @@ def __init__(
         self._curr_dist_to_path = None
         self.safe_dist = safe_dist
 
-        self.custom_rew_dict = custom_rew_dict
+        # load custom reward function from yaml
+        import tools.general as general_utils
```
Comment (ReykCS) [inline rl_utils/rl_utils/utils/reward.py:39]: Immernoch circular imports oder kann das jetzt nach oben?

File: rl_utils/rl_utils/utils/reward.py [hunk 21]
```diff
@@ -37,11 +35,15 @@ def __init__(
...
+
+        self.custom_rew_dict = general_utils.load_rew_fnc(rule)
         self._extended_eval = extended_eval
 
         self.kdtree = None
 
+        # Legacy code
```
Comment (ReykCS) [inline rl_utils/rl_utils/utils/reward.py:46]: Legacy code auch gerne löschen, wenn nicht mehr gebraucht

File: training/tools/env_utils.py (hunk 18 above)
Comment (tuananhroman) [reply training/tools/env_utils.py:107]: Schau ich mir an

File: training/tools/general.py (hunk 19 above)
Comment (tuananhroman) [reply training/tools/general.py:88]: Nope

**Review by ReykCS — APPROVED** [review]
